File: backend/app/schemas/dish.py
from datetime import datetime
from decimal import Decimal
from typing import List, Optional
from pydantic import BaseModel, ConfigDict


class DishIngredientCreate(BaseModel):
    food_item_id: int
    quantity_g: float


class DishCreate(BaseModel):
    name: str
    ingredients: List[DishIngredientCreate]


class DishIngredientRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    food_item_id: int
    food_name: str
    quantity_g: float


class DishRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    user_id: str
    name: str
    name_normalized: str
    total_weight_g: float
    calories_kcal: Optional[float] = None
    protein_g:     Optional[float] = None
    carbs_g:       Optional[float] = None
    fat_g:         Optional[float] = None
    fiber_g:       Optional[float] = None
    sugar_g:       Optional[float] = None
    sodium_mg:     Optional[float] = None
    is_veg:   bool
    is_egg:   bool
    is_vegan: bool
    ingredients: List[DishIngredientRead] = []
    created_at: datetime
    updated_at: datetime


class DishListItem(BaseModel):
    """Lightweight version for list views — no ingredients array."""
    model_config = ConfigDict(from_attributes=True)
    id: int
    name: str
    total_weight_g: float
    calories_kcal: Optional[float] = None
    protein_g:     Optional[float] = None
    carbs_g:       Optional[float] = None
    fat_g:         Optional[float] = None
    is_veg:   bool
    is_egg:   bool
    ingredient_count: int = 0

File: frontend_legacy/src/types/dish.ts
export interface DishIngredient {
  id: number;
  food_item_id: number;
  food_name: string;
  quantity_g: number;
}

export interface CustomDish {
  id: number;
  user_id: string;
  name: string;
  name_normalized: string;
  total_weight_g: number;
  calories_kcal: number | null;
  protein_g:     number | null;
  carbs_g:       number | null;
  fat_g:         number | null;
  fiber_g:       number | null;
  sugar_g:       number | null;
  sodium_mg:     number | null;
  is_veg:   boolean;
  is_egg:   boolean;
  is_vegan: boolean;
  ingredients: DishIngredient[];
  created_at: string;
  updated_at: string;
}

export interface CustomDishListItem {
  id: number;
  name: string;
  total_weight_g: number;
  calories_kcal: number | null;
  protein_g:     number | null;
  carbs_g:       number | null;
  fat_g:         number | null;
  is_veg:   boolean;
  is_egg:   boolean;
  ingredient_count: number;
}

import type { FoodItem } from "@/types/nutrition";

export interface DishIngredientInput {
  food_item_id: number;
  food_name: string;
  quantity_g: number;
  display_amount?: number;
  display_unit?: "g" | "ml" | "qty";
  item_ref?: FoodItem;        // kept in memory for unit conversion, not sent to API
  // nutrition per 100g from food_items (for live preview)
  calories_kcal: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  fiber_g: number;
  serving_size_g: number;
}
